### PR TITLE
Polydataobject abstract

### PR DESCRIPTION
**Pull request by StongeEtienne**
Created an Abstract class for PolyDataObject (AbstractPolyDataObject), for future object such as TactogramObject that will inherit few features of the AbstractPolyDataObject, but does not support texture.

### CODE REVIEW

File: IbisLib/abstractpolydataobject.h [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+public:
+        
+    //static AbstractPolyDataObject * New() { return new AbstractPolyDataObject; }
+    vtkTypeMacro(AbstractPolyDataObject,SceneObject);
+    
+    AbstractPolyDataObject();
+    virtual ~AbstractPolyDataObject();
+    virtual void Serialize( Serializer * ser ) override;
+    virtual void Export() override;
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.h:52]: I would implement Export() and IsExportable() in derived classes, we may have classes that will not be exportable, or that could be exported in different format.

Comment (StongeEtienne) [reply IbisLib/abstractpolydataobject.h]: I moved it and tested the build

File: IbisLib/abstractpolydataobject.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+#include <map>
+#include <QVector>
+#include <QObject>
+#include <vtkProperty.h>
+#include <vtkSmartPointer.h>
+
+class vtkPolyData;
+class vtkTransform;
+class vtkActor;
+class vtkDataSetAlgorithm;
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.h:25]: vtkDataSetAlgorithm isn't used in AbstractPolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+class vtkImageData;
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.h:26]: vtkImageData isn't used in AbstractPolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+public:
+        
+    //static AbstractPolyDataObject * New() { return new AbstractPolyDataObject; }
+    vtkTypeMacro(AbstractPolyDataObject,SceneObject);
+    
+    AbstractPolyDataObject();
+    virtual ~AbstractPolyDataObject();
+    virtual void Serialize( Serializer * ser ) override;
+
+    //virtual void CreateSettingsWidgets( QWidget * parent, QVector <QWidget*> *widgets) = 0;
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.h:53]: Better remove the commented code

File: IbisLib/polydataobject.h [hunk 5]
```diff
@@ -19,6 +19,8 @@ See Copyright.txt or http://ibisneuronav.org/Copyright.html for details.
 #include <vtkProperty.h>
 #include <vtkSmartPointer.h>
 
+#include "abstractpolydataobject.h"
+
 class vtkPolyData;
```
Comment (nist-anka) [inline IbisLib/polydataobject.h:24]: vtkPolyData, vtkTransform, vtkActor, vtkProbeFilter, vtkScalarsToColors, ImageObject, vtkClipPolyData, vtkPassThrough, vtkCutter. vtkPlane and vtkPlanes are not used in polydataobject.h

File: IbisLib/polydataobject.h [hunk 6]
```diff
@@ -32,132 +34,39 @@ class vtkPassThrough;
 class vtkCutter;
 class vtkPlane;
 class vtkPlanes;
+class AbstractPolyDataObject;
```
Comment (nist-anka) [inline IbisLib/polydataobject.h:37]: no need to declare after abstractpolydataobject.h was included

File: IbisLib/abstractpolydataobject.cpp [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+Ibis Neuronav
+Copyright (c) Simon Drouin, Anna Kochanowska, Louis Collins.
+All rights reserved.
+See Copyright.txt or http://ibisneuronav.org/Copyright.html for details.
+
+     This software is distributed WITHOUT ANY WARRANTY; without even
+     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+     PURPOSE.  See the above copyright notice for more information.
+=========================================================================*/
+#include <QFileDialog>
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.cpp:11]: Sorry for being "emmerdeuse" (no English equivalent I know of) I would profit from the occasion to clean the code a little. QFileDialog is not used  neither in AbstractPolyDataObject nor in PolyDataObject. We should not include polydataobject.h here.

File: IbisLib/abstractpolydataobject.cpp [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+#include <QMessageBox>
+#include "polydataobject.h"
+#include <vtkPolyData.h>
+#include <vtkPolyDataMapper.h>
+#include <vtkRenderer.h>
+#include <vtkActor.h>
+#include <vtkTransform.h>
+#include <vtkAssembly.h>
```
Comment (nist-anka) [inline IbisLib/abstractpolydataobject.cpp:19]: vtkRenderer and vtkAssenly are not used in AbstractPolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    vtkScalarsToColors * GetCurrentLut();
+    void InitializeClippingPlanes();
+        
+    vtkPolyData *PolyData;
+    vtkSmartPointer<vtkPassThrough> m_clippingSwitch;
+    vtkSmartPointer<vtkPassThrough> m_colorSwitch;
+
+    int LutIndex;
+    vtkSmartPointer<vtkScalarsToColors> CurrentLut;
+    ImageObject * ScalarSource;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:126]: ScalarSource and related features should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    void SavePolyData( QString &fileName );
+     
+public slots:
+
+    void OnStartCursorInteraction();
+    void OnEndCursorInteraction();
+    void OnCursorPositionChanged();
+    void OnReferenceChanged();
+    void OnScalarSourceDeleted();
+    void OnScalarSourceModified();
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:98]: OnScalarSourceXXX() and related features should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    void SetCrossSectionVisible( bool );
+    void SetClippingEnabled( bool e );
+    bool IsClippingEnabled() { return m_clippingOn; }
+    void SetClippingPlanesOrientation( int plane, bool positive );
+    bool GetClippingPlanesOrientation( int plane );
+
+    bool IsUsingScalarSource();
+    void UseScalarSource( bool use );
+    ImageObject * GetScalarSource() { return this->ScalarSource; }
+    void SetScalarSource( ImageObject * im );
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:87]: ScalarSource and related features should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    vtkSmartPointer<vtkProperty> Property;
+    vtkSmartPointer<vtkProperty> m_2dProperty;
+    
+    typedef std::map< View*,vtkSmartPointer<vtkActor> > PolyDataObjectViewAssociation;
+    PolyDataObjectViewAssociation polydataObjectInstances;
+    
+    int       renderingMode;  // one of VTK_POINTS, VTK_WIREFRAME or VTK_SURFACE
+    int       ScalarsVisible;  // Whether scalars in the PolyData are used to color the object or not.
+    int       VertexColorMode;   // 0 : use scalars in data, 1 : get scalars from object ScalarSourceObjectId
+    int       ScalarSourceObjectId;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:138]: ScalarSource and related features should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    void InitializeClippingPlanes();
+        
+    vtkPolyData *PolyData;
+    vtkSmartPointer<vtkPassThrough> m_clippingSwitch;
+    vtkSmartPointer<vtkPassThrough> m_colorSwitch;
+
+    int LutIndex;
+    vtkSmartPointer<vtkScalarsToColors> CurrentLut;
+    ImageObject * ScalarSource;
+    vtkSmartPointer<vtkScalarsToColors> LutBackup;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:127]: LUT feature should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    virtual void Release( View * view ) override;
+
+    int GetRenderingMode() { return this->renderingMode; }
+    void SetRenderingMode( int mode );
+    void SetScalarsVisible( int use );
+    vtkGetMacro( ScalarsVisible, int );
+    int GetLutIndex() { return LutIndex; }
+    void SetLutIndex( int index );
+    int GetVertexColorMode() { return VertexColorMode; }
+    void SetVertexColorMode( int mode );
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:69]: ScalarProperties, LUT and VertexColoring features should move to PolyDataObject

File: IbisLib/abstractpolydataobject.h [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+#include <vtkSmartPointer.h>
+
+class vtkPolyData;
+class vtkTransform;
+class vtkActor;
+class vtkDataSetAlgorithm;
+class vtkImageData;
+class vtkProbeFilter;
+class vtkScalarsToColors;
+class ImageObject;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.h:29]: Move to PolyDataObject

File: IbisLib/abstractpolydataobject.cpp [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+#include <vtkPolyDataWriter.h>
+#include <vtkImageData.h>
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.cpp:21]: Move to PolyDataObject

File: IbisLib/abstractpolydataobject.cpp [hunk 17]
```diff
@@ -0,0 +1,515 @@
...
+#include <vtkPNGReader.h>
+#include <vtkDoubleArray.h>
+#include <vtkPoints.h>
+
+#include "view.h"
+#include "vtkPiecewiseFunctionLookupTable.h"
+#include "application.h"
+#include "scenemanager.h"
+#include "imageobject.h"
+#include "vtkProbeFilter.h"
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.cpp:31]: Move to PolyDataObject

File: IbisLib/abstractpolydataobject.cpp [hunk 18]
```diff
@@ -0,0 +1,515 @@
...
+        m_cutter[i] = vtkSmartPointer<vtkCutter>::New();
+        m_cutter[i]->SetInputConnection( m_colorSwitch->GetOutputPort() );
+        m_cutter[i]->SetCutFunction( m_cuttingPlane[i]);
+    }
+    m_cuttingPlane[0]->SetNormal( 1.0, 0.0, 0.0 );
+    m_cuttingPlane[1]->SetNormal( 0.0, 1.0, 0.0 );
+    m_cuttingPlane[2]->SetNormal( 0.0, 0.0, 1.0 );
+
+    this->PolyData = 0;
+    this->LutIndex = 0;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.cpp:76]: Move to PolyDataObject

File: IbisLib/abstractpolydataobject.cpp [hunk 19]
```diff
@@ -0,0 +1,515 @@
...
+    this->renderingMode = VTK_SURFACE;
+    this->ScalarsVisible = 0;
+    this->VertexColorMode = 0;
+    this->ScalarSourceObjectId = SceneManager::InvalidId;
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.cpp:80]: Move to PolyDataObject

File: IbisLib/abstractpolydataobject.cpp [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+        clippingPlaneOrientation[0] = GetClippingPlanesOrientation( 0 ) ? 1 : 0;
+        clippingPlaneOrientation[1] = GetClippingPlanesOrientation( 1 ) ? 1 : 0;
+        clippingPlaneOrientation[2] = GetClippingPlanesOrientation( 2 ) ? 1 : 0;
+    }
+    SceneObject::Serialize(ser);
+    ::Serialize( ser, "RenderingMode", this->renderingMode );
+    ::Serialize( ser, "LutIndex", this->LutIndex );
+    ::Serialize( ser, "ScalarsVisible", this->ScalarsVisible );
+    ::Serialize( ser, "VertexColorMode", this->VertexColorMode );
+    ::Serialize( ser, "ScalarSourceObjectId", this->ScalarSourceObjectId );
```
Comment (hgueziri) [inline IbisLib/abstractpolydataobject.cpp:146]: Move to PolyDataObject

**Review by hgueziri — COMMENTED** [review]
We should move all scalar coloring features (VertexColor, LUT, ScalarSource) to PolyDataObject. These are specific properties that are used in the UI of polydata. The abstract class should only keep basic properties of a vtkpolydata

File: IbisLib/abstractpolydataobject.cpp (hunk 8 above)
Comment (StongeEtienne) [reply IbisLib/abstractpolydataobject.cpp:19]: vtkRenderer seems to be used
however vtkAssenly can be removed

**Pull request merged**